beacon-chain/sync/validators_custody.go

```diff
@@ -0,0 +1,147 @@
+package sync
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/v5/async"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/feed"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/feed/state"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/peerdas"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/p2p"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/time/slots"
+)
+
+const updateToAdvertiseCustodyGroupCountPeriod = 1 * time.Minute
+
+func (s *Service) maintainValidatorsCustody() {
+	async.RunEvery(s.ctx, updateToAdvertiseCustodyGroupCountPeriod, s.updateToAdvertiseCustodyGroupCount)
+
+	stateCh := make(chan *feed.Event, 1)
+	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateCh)
+	defer stateSub.Unsubscribe()
+
+	latestProcessedEpoch := params.BeaconConfig().FarFutureEpoch
+
+	for {
+		select {
+		case event := <-stateCh:
+			latestProcessedEpoch = s.handleEvent(event, latestProcessedEpoch)
+		case err := <-stateSub.Err():
+			log.WithError(err).Error("DataColumnSampler1D subscription to state feed failed")
+		case <-s.ctx.Done():
+			log.Debug("Context canceled, exiting data column sampling loop.")
+			return
+		}
+	}
+}
+
+// updateToAdvertiseCustodyGroupCount updates the custody group count to advertise.
+func (s *Service) updateToAdvertiseCustodyGroupCount() {
+	// Retrieve the registered topics, and store them in a map for quick lookup.
+	registeredTopicsSlice := s.subHandler.allTopics()
+	registeredTopics := make(map[string]bool, len(registeredTopicsSlice))
+
+	for _, topic := range registeredTopicsSlice {
+		topicMessage := extractGossipMessage(topic)
+		registeredTopics[topicMessage] = true
+	}
+
+	// Get the node ID.
+	nodeID := s.cfg.p2p.NodeID()
+
+	peerdas.CustodyGroupCountMut.Lock()
+	defer peerdas.CustodyGroupCountMut.Unlock()
+
+	// Get the custody group count.
+	targetCustodyGroupCount := peerdas.TargetCustodyGroupCount.Get()
+
+	// Get the peerDAS info.
+	info, _, err := peerdas.Info(nodeID, targetCustodyGroupCount)
+	if err != nil {
+		log.WithError(err).Error("Failed to get peerDAS info")
+		return
+	}
+
+	for column := range info.CustodyColumns {
+		topicMessage := fmt.Sprintf(p2p.GossipDataColumnSidecarMessage+"_%d", column)
+		if !registeredTopics[topicMessage] {
+			// At least one data column subnet we should be subscribed to is not.
+			return
+		}
+	}
+
+	// All data column subnets we should be subscribed to are.
+	peerdas.ToAdvertiseCustodyGroupCount.Set(targetCustodyGroupCount)
+}
+
+// handleEvent handles a state feed event.
+func (s *Service) handleEvent(event *feed.Event, latestProcessedEpoch primitives.Epoch) primitives.Epoch {
```

Comment:
why not just use the head state here ? If the block is not canonical it might lead to bugs when using it

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/pull/14948/commits/5c03353d8570848b7dce99469eff95fb72523bd4.